src/deploy/AWSDeployer.js

```diff
@@ -272,6 +283,26 @@ export default class AWSDeployer extends BaseDeployer {
       this.log.info(chalk`--: updating existing Lambda function configuration {yellow ${functionName}}`);
       await this._lambda.send(new UpdateFunctionConfigurationCommand(functionConfig));
       await this.checkFunctionReady(baseARN);
+      this.log.info(chalk`--: updating existing Lambda function tags {yellow ${functionName}}`);
+      const { Tags: tagsBeforeUpdate } = await this._lambda.send(new ListTagsCommand({
+        Resource: baseARN,
+      }));
+      // set all the tags in the current configuration
+      await this._lambda.send(new TagResourceCommand({
+        Resource: baseARN,
+        Tags: functionConfig.Tags,
+      }));
+      // then remove any extraneous tags
+      const tagsToRemove = Object.keys(tagsBeforeUpdate || {}).filter(
+        (key) => !(key in functionConfig.Tags),
+      );
+      if (tagsToRemove.length) {
+        await this._lambda.send(new UntagResourceCommand({
+          Resource: baseARN,
+          TagKeys: tagsToRemove,
+        }));
+      }
```

Comment:
- I don't really like this, as it will remove existing (manually added) tags from functions. maybe only support adding/updating tags through hedy? or only delete tags with empty values, eg: `foo=`

- the entire tag handling functions could be guarded with `if (functionConfig.Tags) {` in order to avoid extra API calls (and potentially require higher api permissions)

Reply:
@tripodsan I like the idea of only deleting when there's a blank value. I've incorporated that.

However, I don't follow what you are saying about guarding... `functionConfig.Tags` will *always* be defined (see https://github.com/adobe/helix-deploy/blob/ba3e66603074e892415539fda50cb193a4ef19f3/src/deploy/AWSDeployer.js#L223). While it is obviously possible to check that the values in `functionConfig.Tags` do or do not match the current tags, in the normal course of processing, they will *always* be different since `updated` is set to the current timestamp. So doing an equivalence test would actually require an extra API call (to get the current tag list) which isn't really necessary.

With the updated code, there can be at most 3 API calls:

1. To update the current tags (always necessary at minimum to set `updated`).
2. IF there are tags to remove (i.e. `foo=`), list the current tags.
3. IF there are tags to remove that are currently set, remove them.

The second call could be removed but then the UntagResource API call could be made unnecessarily.

Reply:
ok, now I understand.